src/nodes/chunk_insert_state.c

```diff
@@ -513,6 +513,31 @@ adjust_projections(ChunkInsertState *cis, ChunkDispatch *dispatch, Oid rowtype)
 	}
 }
 
+/* Assumption: we have acquired a lock on the chunk table. This is
+ *      important because lock acquisition order is always orignal chunk,
+ *      followed by compressed chunk to prevent deadlocks.
+ * We now try to acquire a lock on the compressed chunk, if one exists.
+ * Note that the insert could have been blocked by a recompress_chunk operation.
+ * So the compressed chunk could have moved under us. We need to refetch the chunk
+ * to get the correct compressed chunk id (github issue 3400)
+ */
+static Relation
+lock_associated_compressed_chunk(int32 chunk_id, bool *has_compressed_chunk)
+{
```

Comment:
You have the `chunk` already in the caller. Why are you not passing that to the function instead of fetching it again?

Reply:
See explanation above.